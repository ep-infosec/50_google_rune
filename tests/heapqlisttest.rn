//  Copyright 2021 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

class A(self) {
  final(self) {
    println "Destroying A ", <u32>self
  }
}

class B(self, a: A, cost) {
  self.cost = cost
  a.pushB(self)
  final(self) {
    println "Destroying B ", <u32>self
  }

  operator < (b1: B, b2: B) {
    return b1.cost < b2.cost
  }
}

class C(self, a: A, cost) {
  self.cost = cost
  a.pushC(self)
  final(self) {
    println "Destroying C ", <u32>self
  }

  operator > (c1: C, c2: C) {
    return c1.cost > c2.cost
  }
}

relation HeapqList A B cascade
relation HeapqList A C cascade (false)

list = [83, 1, 8, 123, 929, 33, 0, 10]
a = A()
for i in range(list.length()) {
  B(a, list[i])
  C(a, list[i])
}
do {
  b = a.popB()
} while !isnull(b) {
  println "popped B ", b.cost
  b.destroy()
}
do {
  c = a.popC()
} while !isnull(c) {
  println "popped C ", c.cost
  c.destroy()
}
for i in range(list.length()) {
  B(a, list[i])
}
a = null(a)
