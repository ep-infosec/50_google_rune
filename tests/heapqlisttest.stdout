popped B 0
Destroying B 6
popped B 1
Destroying B 1
popped B 8
Destroying B 2
popped B 10
Destroying B 7
popped B 33
Destroying B 5
popped B 83
Destroying B 0
popped B 123
Destroying B 3
popped B 929
Destroying B 4
popped C 929
Destroying C 4
popped C 123
Destroying C 3
popped C 83
Destroying C 0
popped C 33
Destroying C 5
popped C 10
Destroying C 7
popped C 8
Destroying C 2
popped C 1
Destroying C 1
popped C 0
Destroying C 6
Destroying A 0
Destroying B 1
Destroying B 3
Destroying B 0
Destroying B 6
Destroying B 2
Destroying B 4
Destroying B 5
Destroying B 7
